### ISSUE TITLE
Higher note snaps in the editor

### ISSUE THREAD

**Issue by 1033Forest** [stepmania/stepmania#1110]
A feature from SM AMX. I think higher note snaps (up to 1536th) should be added. That way I can test how the new noteskin system picks colors for any note over 192nd. Plus with the new noteskin system, there is finally a way to not let the existing noteskins break, as the new noteskin system is not limited to 192nd quantization. This feature will be for 5.1, not 5.0.

If they were added these would be the note snaps (16, instead of just 9):
4th, 8th, 12th, 16th, 24th, 32nd, 48th, 64th, 96th, 128th, 192nd, 256th, 384th, 512th, 768th, 1536th

**Comment by dguzek** [comment]
> I think higher note snaps (up to 1536th) should be added.

Why do you think this feature should be added?

**Comment by 1033Forest** [comment]
Just to test out the new noteskin system that has been added in 5.1, since it's not limited to just 192nd quantization now.
And if there are 192nd notes, where's the snap for 96th notes?

**Comment by sillybear** [comment]
Honestly, that would be a lot of inconvenience. I don't see any application for using extra unnecessary note snaps. This would mean adding missing note beats in noteskins too.

**Comment by 1033Forest** [comment]
Then... I don't know what's the point of adding them in SM AMX then. But I've seen a 128th note used in a pump file for Arch of Darkness...
https://www.youtube.com/watch?v=rDS7ZhwQOkI 0:44 See that pastel blue center note? That's a 128th note

**Comment by sillybear** [comment]
I would suggest a something like a Beat Snap Options so you can toggle beats (4th-1536th) on and off to be used in editor.

**Comment by 1033Forest** [comment]
Hmm... that actually sounds like a better idea. I could enable the higher beats just to test out the new noteskin system.

The SM 3.9 AMX note colors for notes higher than 64th:
96th - Pastel red/pink (#FF9999)
128th - Pastel blue (#9999FF)
192nd - Pastel green (#99FF99)
256th - Maroon (#993300)
384th - Extremely dark, kinda blueish green color (#003333)
768th - Gray (same as 192nds on a normal SM build) (#999999)

If 1536ths were added:
512th - Dark navy blue (#000033)
1536th - Black (#000000)

**Comment by wolfman2000** [comment]
I need a reminder of what songs have non-common music time so that I can experiment with non quantized notes again. I personally want to mess with 5th, 10th, and 20th notes for a 5/4 song since I tried it on an old StepMix song and forgot about it.

...I mean, I guess I could just use another Go (EK/Mahalo) file...but if you guys know of something better, I'm open.

**Comment by 1033Forest** [comment]
Another reason why I wanted the feature of higher note quantizations, was to experiment with them. 

In the zenbuquan noteskin there are notes 5x16. Using this feature, I would be able to see how the notes look like after the 8th row in that file.

**Comment by 1033Forest** [comment]
Just calculated, found out: a 1536th note has a quantization of 0.00260416666666 (1/384). If the noteskin system with 8 colors is still used, what color would a note like that be?

**Comment by kyzentun** [comment]
The notedata is really hardcoded to 48 parts per beat.  There are probably hidden things all over the engine that will break if we change it.

A comment in zenbuquan notes that normal 12ths came out the wrong color when I had tons of quantizations.  So the algorithm used to quantize notes probably needs to be rewritten.

Rather than kludging in finer quantizations, it would be better to invent a note data format that doesn't have fixed quantization limits.  Something to worry about for SM6.

**Comment by shakesoda** [comment]
This is pretty much just not a good idea, what you want is a file format with millisecond timing.

**Comment by Kaiveran** [comment]
There's a lot of things connected with this and I'm just dumping my thoughts/concerns.

What's interesting/frustrating to me is that while the internal note data may be limited, we already have a _simfile format_ that doesn't have a fixed quantization limit, which is SM/SSC. You can (or at least could at one time) use any odd number of rows per measure in the file, and they will be understood by the parser. I used this trick back in the day for some songs with odd tuplets. The notes just got herded to the nearest 192nd. Of course, I understand there were several factors keeping the note data from changing to match this power, but basically, the only reason one would _need_ to be limited to 192nds these days is to maintain compatibility with DWI, which is completely useless these days unless you're stepping for other dinosaurs like FFR.

Of course, writing notes directly to times has its appeal, for it's simplicity and purity. But beats are what literally everyone is used to. Beats are the foundation of all the music we make, and the way we color/differentiate those symbols. So I'd have to disagree with @shakesoda here that a simfile format in which notes are encoded to milliseconds is ideal. Rather, I think notes should still be encoded to fractions of beats, and just snapped to the nearest millisecond in the program. This way, even loads of quantizations will be no problem, as SM will actually know what exact beat is intended to be used and the problem of slight timing error + coarser timing resolution at high BPMs won't get in the way of accuracy in the NewSkin system.

I think this would necessitate a new way of placing notes in editor. Something dynamic and user-controlled, like being able to save different tuplet values on your profile, and be able to "write in" a beat value if needed. I'll explore this in a more relevant issue. (Also I should really finish up my sim format idea.)

But to bring things back to reality, 192nds are what we have to work with right now, so that's what we have to consider even in the new systems being designed. @kyzentun, I'm concerned about the consequences that your experience designing zenbuquan might have on my NewSkin in progress. Even to program minimal quantizations for 5/7/9-tuplets in it requires a ludicrous **5040** parts_per_beat.
(See here: https://github.com/Kaiveran/babykai/blob/master/NewSkins/babykai/notes.lua#L687-706)

Would the zenbuquan error stick its cock up my quantizations too, even though they're spread out over 2 beats (quantum_time = 2)?

**Comment by shakesoda** [comment]
the SM format gets increasingly less efficient as you use absurd note snaps, and more importantly, so does making noteskins.

**Comment by shakesoda** [comment]
so yes, you want millisecond timing after some point in the format, and for the editor to just abstract it for you.

**Comment by Kaiveran** [comment]
**Or**, we could not have the concept of measures canonized in the simfile format at all. I get where you're coming from, but you've made me think _we don't need to play by SM's rules_. What if we offloaded the abstraction _totally_ to the editor and simply encoded notes sequentially by beat as they came? Imagine a notechart looking something like this:

Left - 1 - Hold Start
Left - 1.5 - Hold End
Down - 2 - Tap
Up - 2.2 - Tap
Down - 2.4 - Tap
Up - 2.6 - Tap
Down - 2.8 - Tap
Up - 3 - Tap
Up - 3.5 - Lift
Up - 3.75 - Tap
Right - 4 - Tap
Left - 4+1/7 - Tap
Right - 4+2/7 - Tap
Left - 4+3/7 - Tap
Right - 4+4/7 - Tap
Left - 4+5/7 - Tap
Right - 4+6/7 - Tap


Look at this. 8ths, 16th, 20ths, and 28ths all in the same measure. In the SM/SSC method you would need **560** lines to encode this measure. But do it this way, and you only need **17**...the same as the number of notes!

...um...

Sorry, I get over excited sometimes. I don't mean to be an uppity little bitch. I know you have more knowledge and have been at this for much longer than me. To accomplish what I'm thinking about - new note data structure and a totally new chart format to parse into that structure - is probably way more difficult than I can understand and intersects with a bunch of other shit that'll break. I just look at all this new stuff and the more I think about it, the more I feel like something is fundamentally limiting about the way things are handled currently. Hence I'm compelled to ask the more "upstream" questions of do we need to be locked into things like measures and stuff in the first place, when it comes to actually encoding the notes. Maybe the ideal is to have just have one beat-addressed line per note, like above...no more zeros.

But yeah, I'll STFU about that specifically until I can publish a viable, complete alternative along these lines. (I just destroyed my previous new format idea by thinking of this, oh well)

**Comment by 1033Forest** [comment]
http://i.imgur.com/rINzV7j.png
Here's a color scheme I thought of for higher notes. The first 8 are from SM 3.9. 96th-768th (except 512th) are from AMX. 512th and everything after that, I thought of the colors.

**Comment by Kaiveran** [comment]
@1033Forest, this really isn't the venue to discuss noteskin design. That should be done on the forums. Here (I think) we're chiefly concerned with the technical details of implementing non-fixed quantizations and encoding charts for them.

@sillybear, we can discuss more how a future editor should work and implement these capabilities in #1015 .

**Comment by 1033Forest** [comment]
I edited a .sm file to attempt to get 384ths to work, but they didn't. What my test file did made 192nd hands instead.
My file: https://www.mediafire.com/?hedsrds7v53dbh6
If you load this file in AMX, very dark greenish blue arrows will appear. These are the 384th notes.
IDK what happens if you try to load this in older SM versions. Maybe the 384ths might appear, but they might all be green.

**Comment by 1033Forest** [comment]
I think this feature would better off be in SM5.1+ or SM6.
In the options menu, there would be "Beat Snap Options". In there you can toggle which beat snaps appear in the editor while navigating through them (4th-192nd is enabled by default but you can enable up to 1536th).
The note snaps that can be enabled and disabled would be (in order):
4th, 8th, 12th, 16th, 24th, 32nd, 48th, 64th, 96th, 128th, 192nd, 256th, 384th, 512th, 768th, 1536th.
128th, 256th, 384th, 512th, 768th, and 1536th would be disabled by default.

Noteskins could still be made using the 8 color format, just all the notes after 64th would be colored depending on what beat/frame they are on.

**Comment by Kaiveran** [comment]
@1033Forest I'm actually thinking something broader in scope than just multiplying the current tuplets, since we're due to leave fixed quantizations entirely, we need systems with greater user control that can take that into account. See here to continue the discussion: https://github.com/stepmania/stepmania/issues/1015#issuecomment-220454416

**Comment by 1033Forest** [comment]
In a future 5.1, I am planning to implement 128th notes and see if they work. They are the notes between 96ths and 192nds. The only quantization beyond 192nd to add is 384th. Even though 384ths are useless and will probably not be used in any file at all, they are required for 128ths to work properly.

Since the note data is hardcoded to 48 rows per beat, I tried my 384th test file and instead of giving me 128ths and 384ths, it just converted most of the notes into a bunch of hands. I could try changing it to 96 to see what happens.

Not bothering to add 256ths/768ths for now since these notes will probably never be used in any simfile ever. I saw one use of a 128th in a PIU edit, as I said in an earlier comment to this.
